api/external_auth.proto

```diff
@@ -0,0 +1,151 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2;
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with OK or NOT OK.
+service Authorization {
+  rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+  // The name of the tenant for multi-tenant support.
+  string tenant_name = 1;
+
+  // The request attributes.
+  AttributeContext attributes = 2;
+}
+
+message CheckResponse {
+  // Status OK allows the request. Any other response indicates the request should be denied.
+  google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// The fields follow the Istio attribute vocabulary.
+// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+// The field type becomes the attribute type, and the field path becomes the attribute name.
+// For example, the attribute `source.ip` maps to field `AttributeContext.source.ip`.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The IP address of the peer.
+    string ip = 1;
+
+    // The network port of the peer.
+    int64 port = 2;
+
+    // The canonical service name of the peer.
+    string service = 3;
+
+    // The labels associated with the peer.
+    // These could be pod labels for Kubernetes or tags for VMs.
+    // The source of the labels could be an X.509 certificate or other configuration.
+    map<string, string> labels = 4;
+
+    // The validated identity of this peer.
+    // For example, the identity associated with the workload such as a service account.
+    // If an X.509 certificate is used to assert the identity this field should be sourced from
+    // `Subject` and `Subject Alternative Names`.
+    repeated string principals = 5;
+
+    // Other claims about the peer that are not captured in labels and principals.
+    google.protobuf.Struct claims = 6;
```

Comment:
I think this should be metadata, https://github.com/envoyproxy/data-plane-api/blob/master/api/base.proto#L57. Or at least, there should be the capability of passing metadata expressed in Envoy's configs to the auth service (as we do in logging, see https://github.com/envoyproxy/data-plane-api/blob/master/api/filter/accesslog/accesslog.proto#L65).

Reply:
claims is the information from auth (thinking JWT), while labels and metadata are associated with the source. For now, we should limit the number of attributes people can use for policies. Even with multi-layer security system, you shouldn't need too many attributes.

Reply:
These claims are about the peer, so will in general be different on every connection.  If filter metadata is required it should be on the `AttributeContext`

Reply:
That is not true at Google. We can do service-to-service equivalence of mTLS, so when a client sends traffic to load balanced servers. The metadata can be shared across connection. Otherwise, mTLS will be useless for load balanced services, since you have to restart mTLS hand shake all the time.

Reply:
We need to find a place for filter config params in `AttributeContext`.  These are peer claims.